### PR TITLE
Refactor DartDocConfig object usage, part 1

### PR DESCRIPTION
**Pull request by jcollins-g**
A first step toward making dartdoc_options.yaml play nicely with command line options.  The way dartdoc handles configuration right now is very confusing -- some things are referenced in a global configuration object, some things are referenced via an object that's passed around, some things are referenced directly from  parameters that are also in the config object...  On top of that, Dartdoc's internal caching often had to be bypassed for testing because cached values were dependent on configuration objects.

Unfortunately, adding the dartdoc options file added Yet Another Way of configuring things and didn't tie it in to the old ways.  To integrate that, we first need to get the rest of dartdoc's configuration under control.

This does that. There's only one way to look at each of the major flags now -- via a config object that is passed in to DartDoc objects at construction time.  This makes it safe to, among other things, have multiple DartDoc objects with different configuration objects generate docs inside the same program (a pretty basic prerequisite for #1609).

Also:
*  Sorts parameters to the configuration object so it is easier to tell visually that things match up
*  Eliminates the global config getter, setConfig(), and so on
*  Revives a commented out test that verifies exclude flag behavior
*  Refactors LibraryContainer and subclasses so their sorting can be tested outside of actually constructing an enormous package, because the old way of avoiding this was to leverage global configuration objects.

Some of the command line options (those that involve setting up the generators) aren't moved in this PR, but will be in a followup as that's a fairly large change too.

### CODE REVIEW

File: lib/src/model.dart [hunk 1]
```diff
@@ -4596,59 +4588,79 @@ class PackageGraph extends Canonicalization with Nameable, Warnable {
...
-  String _name;
 
-  Category(this._name, this.package, this.dartdocOptions);
+  /// Does this container represent the SDK?  This can be false for containers
+  /// that only represent a part of the SDK.
+  bool get isSdk => false;
 
-  @override
-  String get name => _name;
+  // Order by which this container should be sorted.
```
Comment (pq) [inline lib/src/model.dart]: => /// ?

File: bin/dartdoc.dart [hunk 2]
```diff
@@ -252,26 +248,30 @@ main(List<String> arguments) async {
...
-      reexportMinConfidence:
-          double.parse(args['ambiguous-reexport-scorer-min-confidence']),
-      verboseWarnings: args['verbose-warnings'],
-      excludePackages: args['exclude-packages'],
-      dropTextFrom: dropTextFrom,
-      validateLinks: args['validate-links']);
-
-  DartDoc dartdoc = new DartDoc(inputDir, excludeLibraries, sdkDir, generators,
-      outputDir, packageMeta, includeLibraries, includeExternals);
+  DartDocConfig config = new DartDocConfig.fromParameters(
```
Comment (kevmoo) [inline bin/dartdoc.dart]: @jcollins-g – FYI https://pub.dartlang.org/packages/build_cli 😄

File: lib/src/model.dart (hunk 1 above)
Comment (jcollins-g) [reply lib/src/model.dart]: done

File: bin/dartdoc.dart (hunk 2 above)
Comment (jcollins-g) [reply bin/dartdoc.dart]: @kevmoo  Hmmm. What I'm really looking for is a non-throwing, tolerant version of safe_config thrown in a blender with build_cli and a sprinkle of the analyzer's directory-specific configuration semantics.  I have that a few PRs out on this line of development.

**Pull request merged**